test/mpi/test_mpi_cuda.c

```diff
@@ -0,0 +1,668 @@
+    alloc_mem = allocator->alloc(1);
+    if (alloc_mem.ptr == 0) {
+        return 0;
+    }
+
+    allocator->free(&alloc_mem);
+    pop_and_release_primary_context(cu_dev);
+
+    return 1;
+}
+
+int main(int argc, char **argv)
+{
+#if ENABLE_MT
+    int required = MPI_THREAD_MULTIPLE;
+#else
+    int required = MPI_THREAD_SINGLE;
+#endif
+    int provided;
+    int comm_size;
+    int rank;
+    int dev_count, dev_idx;
+    CUdevice cu_dev;
+    CUcontext cu_ctx;
+    void *gold_data;
+    int i, j, k, l;
+    const test_t *test;
+    const allocator_t *allocator_send, *allocator_recv;
+    int res;
+
+    MPI_Init_thread(&argc, &argv, required, &provided);
+    MPI_Comm_size(MPI_COMM_WORLD, &comm_size);
+    if (comm_size != 2) {
+        PRINT_ROOT("This test requires 2 processes, not %d\n", comm_size);
+        goto out;
+    }
+
+    MPI_Comm_rank(MPI_COMM_WORLD, &rank);
+
+    CUDA_CALL(cuInit(0));
+    CUDA_CALL(cuDeviceGetCount(&dev_count));
+    if (dev_count < 1) {
+        PRINT_ROOT("This test requires at least 1 GPU\n");
+        goto out;
+    }
+
+    for (dev_idx = dev_count - 1; dev_idx > -1; --dev_idx) {
+        CUDA_CALL(cuDeviceGet(&cu_dev, dev_idx));
+        CUDA_CALL(cuDevicePrimaryCtxRetain(&cu_ctx, cu_dev));
+    }
+
+    if ((rank == 1) && (dev_count > 1)) {
+        CUDA_CALL(cuDeviceGet(&cu_dev, 1));
+    }
+
+    gold_data = create_gold_data(SIZE_L);
+
+    ucs_log_push_handler(ucx_errors_and_warnings_counter);
+
+    for (i = 0; i < array_size(tests); ++i) {
+        test = tests + i;
+        PRINT_ROOT("\nTEST[%d]: %s\n", i, test->description);
+
+        if (dev_count < test->min_gpus) {
+            PRINT_ROOT(TEST_NAME_FMT ": SKIP (min %d GPUs needed)\n", i, "*",
+                       "*", test->min_gpus);
+            continue;
+        }
+
+        if (test->mthread_support && (provided != MPI_THREAD_MULTIPLE)) {
+            PRINT_ROOT(TEST_NAME_FMT ": SKIP (multi-thread is not provided)\n",
+                       i, "*", "*");
+            continue;
+        }
+
+        for (j = 0; j < array_size(allocators); ++j) {
+            allocator_send = allocators + j;
+            if (!check_allocator(allocator_send, cu_dev)) {
+                PRINT_ROOT(TEST_NAME_FMT ": SKIP (not supported)\n", i,
+                           allocator_send->name, "*");
+                continue;
+            }
+
+            for (k = 0; k < array_size(allocators); ++k) {
+                allocator_recv = allocators + k;
+                if (!check_allocator(allocator_recv, cu_dev)) {
+                    PRINT_ROOT(TEST_NAME_FMT ": SKIP (not supported)\n", i,
+                               allocator_send->name, allocator_recv->name);
+                    continue;
+                }
+
+                for (l = 0; l < array_size(sizes); ++l) {
+                    total_ucx_errors_and_warnings = 0;
+                    res = test->func(allocator_send, allocator_recv, gold_data,
```

Comment:
why are we using gold_data of size L for all sizes?

Reply:
Because gold_data of size S is equal to first S bytes of gold_data of size M, and gold_data of size M is equal to first M bytes of gold_data of size L.

Reply:
right, i somehow thought there is larger value